File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/chain_of_responsibility/solution/CustomerServiceRepresentative.m
% CustomerServiceRepresentative.m

classdef CustomerServiceRepresentative < Handler
	methods
		% constructor:
		function theCustomerServiceRepresentative = CustomerServiceRepresentative(aHandler)
			theCustomerServiceRepresentative@Handler(aHandler);
		end
	end

	methods (Access = protected)
		function result = canHandle(theCustomerServiceRepresentative, aCustomerRequestsDiscountEvent)
			result = (aCustomerRequestsDiscountEvent.Percent < 10);
		end

		function doHandle(theCustomerServiceRepresentative, aCustomerRequestsDiscountEvent)
			fprintf('I am a CustomerServiceRepresentative.\n');
			fprintf('I can give you a %d%% discount.\n', aCustomerRequestsDiscountEvent.Percent);
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/chain_of_responsibility/solution/CustomerServiceRepresentative.java
// CustomerServiceRepresentative.java

public class CustomerServiceRepresentative extends Handler
{
	public CustomerServiceRepresentative(Handler aSuccessor)
	{
		super(aSuccessor);
	}

	protected boolean canHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent)
	{
		return (aCustomerRequestsDiscountEvent.getPercent() < 10);
	}

	protected void doHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent)
	{
		System.out.println("I am a CustomerServiceRepresentative.");
		System.out.println("I can give you a " + aCustomerRequestsDiscountEvent.getPercent() + "% discount.");
	}
}
